itglue/client.go

```diff
@@ -0,0 +1,230 @@
+		return fmt.Errorf("client_options: %v", err)
+	}
+	if c.Token == "" {
+		return errors.New("missing token")
+	}
+	if c.Endpoint == "" {
+		return errors.New("missing endpoint")
+	}
+	_, ok := URL[c.Endpoint]
+	if !strings.HasPrefix(c.Endpoint, "https://") && !ok {
+		return fmt.Errorf("invalid endpoint, not https or in %v", URL)
+	}
+	return nil
+}
+
+func NewITGlueAdapter(conf ITGlueConfig) (*ITGlueAdapter, chan struct{}, error) {
+	var err error
+	a := &ITGlueAdapter{
+		conf:   conf,
+		ctx:    context.Background(),
+		doStop: utils.NewEvent(),
+	}
+
+	if strings.HasPrefix(conf.Endpoint, "https://") {
+		a.endpoint = conf.Endpoint
+	} else if v, ok := URL[conf.Endpoint]; ok {
+		a.endpoint = v
+	} else {
+		return nil, nil, fmt.Errorf("not a valid api endpoint: %s", conf.Endpoint)
+	}
+
+	a.uspClient, err = uspclient.NewClient(conf.ClientOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	a.httpClient = &http.Client{
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 10 * time.Second,
+			}).Dial,
+		},
+	}
+
+	a.chStopped = make(chan struct{})
+
+	a.wgSenders.Add(1)
+	go a.fetchEvents(logsURL)
+
+	go func() {
+		a.wgSenders.Wait()
+		close(a.chStopped)
+	}()
+
+	return a, a.chStopped, nil
+}
+
+func (a *ITGlueAdapter) Close() error {
+	a.conf.ClientOptions.DebugLog("closing")
+	a.doStop.Set()
+	a.wgSenders.Wait()
+	err1 := a.uspClient.Drain(1 * time.Minute)
+	_, err2 := a.uspClient.Close()
+	a.httpClient.CloseIdleConnections()
+
+	if err1 != nil {
+		return err1
+	}
+
+	return err2
+}
+
+func (a *ITGlueAdapter) fetchEvents(url string) {
+	defer a.wgSenders.Done()
+	defer a.conf.ClientOptions.DebugLog(fmt.Sprintf("fetching of %s events exiting", url))
+
+	lastCursor := ""
+	for !a.doStop.WaitFor(30 * time.Second) {
+		// The makeOneRequest function handles error
+		// handling and fatal error handling.
+		items, newCursor := a.makeOneRequest(url, lastCursor)
+		lastCursor = newCursor
+		if items == nil {
+			continue
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err == nil {
+					continue
+				}
+				a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+				a.doStop.Set()
+				return
+			}
+		}
+	}
+}
+
+func (a *ITGlueAdapter) makeOneRequest(url string, lastCursor string) ([]utils.Dict, string) {
+
+	// Prepare the request body.
+	reqData := opRequest{}
+	b, err := json.Marshal(reqData)
+	if err != nil {
+		a.doStop.Set()
+		return nil, ""
+	}
+
+	// Get request timestamp
+	currentTime := time.Now().UTC()
+	timeFormat := "2006-01-02T15:04:05.000000Z07:00"
+	thirtySecondsAgo := currentTime.Add(-30 * time.Second)
```

Comment:
This could potentially introduce a small blind spot, for example:
Let's say you do your first fetch, you get 10 events and process them.
Processing those 10 events take 2 seconds.
Then you sleep 30 seconds, and poll the API again.
When you query the API, you'll do Now-30s, but this will have a 2s gap between the last time you queried the API and this new query.

I don't know the details of this API (like if they support "infinite" cursors, which makes this easier), but one way to do it could be:
- query the API
- process the events returned
- record the last timestamp of the last event returned in a variable
- sleep 30s
- make the next API call, and use the last timestamp returned from the previous query as the `[created_at]` filter.

In theory this would always make sure you don't get any gaps. See what I mean?

Reply:
whenever i called the API with a time.now timestamp, i didn't get anything. if i waited 30s, i would get the data from the (formerly) time.now timestamp. so i added a buffer. 

the API is pretty cut and dry, if no data, no cursor, if data, "cursor" appends the timestamp of the next page to the filter param in the URL and that's saved as the cursor param which gets used as the URL it calls next

Reply:
Ok sounds good.